### PR TITLE
WIP: Share include processor result between task within a build

### PR DESCRIPTION
**Pull request by lacasseio**
### Context
See: https://github.com/gradle/gradle-native/issues/302

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and [CI build status](https://builds.gradle.org/project.html?projectId=Gradle_Check&tab=projectOverview&branch_Gradle_Check=lacasseio%2Fnative%2Fsome-work)
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/DefaultSourceIncludesSearchPath.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import org.gradle.api.internal.changedetection.state.FileSystemSnapshotter;
+import org.gradle.internal.file.FileType;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class DefaultSourceIncludesSearchPath implements SourceIncludesSearchPath {
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/DefaultSourceIncludesSearchPath.java]: Not sure if this is terribly better:
`SourceIncludeFileResolver`
`DefaultSourceIncludeFileResolver`

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/DefaultSourceIncludesSearchPath.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        }
+        List<File> quotedSearchPath = new ArrayList<File>(includePaths.size() + 1);
+        quotedSearchPath.add(sourceDir);
+        quotedSearchPath.addAll(includePaths);
+        return quotedSearchPath;
+    }
+
+    @Override
+    public void searchForDependency(String include, SearchResult dependencies) {
+        searchForDependency(includePaths, include, dependencies, false);
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/DefaultSourceIncludesSearchPath.java]: Can't this just return the SearchResult?

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/DefaultSourceIncludesSearchPath.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    private final Map<File, Map<String, IncludeFileImpl>> includeRoots;
+
+    public DefaultSourceIncludesSearchPath(List<File> includePaths, FileSystemSnapshotter fileSystemSnapshotter) {
+        this.includePaths = includePaths;
+        this.fileSystemSnapshotter = fileSystemSnapshotter;
+        this.includeRoots = new HashMap<File, Map<String, IncludeFileImpl>>();
+    }
+
+    @Override
+    public SourceIncludesSearchPath asQuotedSearchPath(File sourceFile) {
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/DefaultSourceIncludesSearchPath.java]: I would just make this simpler and pass the extra information to searchForDependency vs building a specialized anonymous class here.

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/DefaultSourceIncludesResolver.java [hunk 4]
```diff
@@ -356,11 +269,12 @@ public int hashCode() {
         void visitUnresolved();
     }
 
-    private static class BuildableResult implements IncludeResolutionResult {
+    private static class BuildableResult implements IncludeResolutionResult, DefaultSourceIncludesSearchPath.SearchResult {
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/DefaultSourceIncludesResolver.java]: I think if we just return the SearchResult, we don't have to implement this here.

File: subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/IncrementalCompileFilesFactory.java [hunk 5]
```diff
@@ -186,7 +246,7 @@ private FileVisitResult visitFile(File file, FileSnapshot fileSnapshot, Collecti
     /**
      * Details of a file that are independent of where the file appears in the file include graph.
      */
-    private static class FileDetails {
+    public static class FileDetails {
```
Comment (big-guy) [inline subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/incremental/IncrementalCompileFilesFactory.java]: Be sure to check if we can reduce the visibility on this again

**Comment by lacasseio** [comment]
While trying to get a test exposing a problem with some spike I did I found out that `IncludeFileImpl` wasn't giving the right result during `equals`. Fixing the issue did improve the performance, but it's still a bit slower than what is in master (about 1s slower).

**Comment by blindpirate** [comment]
Are we still going to make it in `4.7RC1`? If not, please remove the milestone.

**Comment by blindpirate** [comment]
Removed `4.7RC1` since @lacasseio is off this week.

**Comment by big-guy** [comment]
Thanks @blindpirate. This is parked for now since it doesn't show the improvement we wanted.

**Comment by lacasseio** [comment]
Other work superseded this work, closing now.

**Pull request closed without merging**